### ISSUE TITLE
2023/01/25 ~ 2023/02/02

### ISSUE THREAD

**Issue by njs03332** [njs03332/ml_study#56]
- 1: 11.3.5
- 2: 11.3.6
- 3: 11.4.1~11.4.2
- 2/2 목요일 10:00pm

<img width="133" alt="스크린샷 2023-01-25 오후 10 35 03" src="https://user-images.githubusercontent.com/37107740/214577152-9da480cd-c035-4347-8ef1-a13163578823.png">

**Comment by givitallugot** [comment]
## 11.3.5 Adam과 Nadam 최적화

#### Adam
- **Adam**: Adaptive Moment Estimation(적응적 모멘트 추정)으로 모멘텀 최적화(지난 그레이디언트의 지수 감소 평균)와 RMSProp(지난 그레이디언트 제곱의 지수 감소된 평균)의 아이디어를 합침
<img src=https://user-images.githubusercontent.com/50584633/216305298-3bcfc6c6-5828-4055-b77c-c8dad073b76b.png>
- 모멘텀 감쇠 파라미터 β1은 보통 0.9로 초기화하고 스케일 감쇠 하이퍼파라미터 β2는 0.999로 초기화하는 경우가 많음

```python
optimizer = keras.optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999)
```
- Adam이 AdaGrad나 RMSProp처럼 적응적 학습률 알고리즘이기 때문에 학습률 하이퍼파라미터 η=0.001을 튜닝할 필요가 적음

#### AdaMax
- Adam에서 단계2를 s <- max(βs, ΔJ(θ))로 바꾸고 단계 4를 삭제, 단계 5에서 s에 비례하여 그레이디언트 업데이트의 스케일을 낮춤 (시간에 따라 감쇠된 그레이디언트의 최댓값)
- 실전에서 AdaMax가 Adam보다 더 안정적이나 일반적으로 Adam의 성능이 더 나음
- Adam이 잘 동작하지 않을 때 시도할 수 있는 옵티마이저 중 하나


#### Nadam
- Adam 옵티마이저에 네스테로프 기법을 더한 것, Adam보다 조금 더 빠르게 수렴
- Adam보다 조금 더 빠르게 수렴, 일반적으로 Adam보다 성능이 좋았지만 PMSProp이 더 나을 때 존재


#### 지금까지 논의한 모든 최적화 기법은 1차 편미분(Jacobian)에만 의존, 최적화 이론에는 2차 편미분(Hessian)을 기반으로 뛰어난 알고리즘도 존재하나 심층 신경망 적용이 어려움, 하나의 출력마다 n개의 1차 편미분이 아니라 n 제곱개의 2차 편미분을 계산해야 하기 때문

#### 희소 모델 훈련
- 모든 최적화 알고리즘은 대부분 파라미터가 0이 아닌 밀집 모델을 만듦
- 빠른 속도, 적은 메모리 모델이 필요하면 희소 모델을 만들어가면서 훈련 (훈련하는 동안 l1 규제를 강하게 적용하면 가능한 많은 가중치를 0으로 만들도록 강제하기 때문에)

**Comment by danbi5228** [comment]
## 11.4 규제를 사용해 과대적합 피하기
- 심층 신경망은 전형적으로 수만 개, 때로는 수백만 개의 파라미터를 가지고 있어 자유도가 매우 높음
  - 그래서 대규모의 복잡한 데이터셋을 학습할 수 있지만, 훈련 세트에 과대적합되기 쉬울 수 있음
- 10장에서 조기종료 구현한 것과 배치 정규화도 규제 방법으로 사용할 수 있음

### 11.4.1 L1과 L2 규제
- 신경망의 연결 가중치를 제한하기 위해 L2규제를 사용하거나, 많은 가중치가 0인 희소 모델을 만들기 위해 L1 규제를 사용할 수 있음
```python

## 규제 강도 0.01을 사용하여 L2 규제를 적용하는 방법
layer = keras.layers.Dense(100, activation="elu", kernel_initializer="he_normal", 
            kernel_regularizer=keras.regularizer.l2(0.01))

## L1과 L2가 모두 필요하면 keras.regularizers.l1_l2() 사용. l1(), l2(), l1_l2() 모두 기본값은 0.01

## 일반적으로 네트워크의 모든 은닉층에 동일한 활성화 함수/초기화 전략/규제 를 적용하므로
## 높은 가독성을 위해 functools.partial() 함수를 이용할 수 있음

from functools import partial
RegularizedDense = partial(keras.layers.Dense, activation="elu", kernel_initializer="he_normal", 
            kernel_regularizer=keras.regularizer.l2(0.01))

model = keras.models.Sequential([
   keras.layers.Flatten(input_shape=[28, 28]),
   RegularizedDense(300),
   RegularizedDense(100),
   RegularizedDense(10, activation="softmax", kernel_intializer="glorot_uniform")
])
```


### 11.4.2 드롭아웃
- 매 훈련 스텝에서 각 뉴런(입력 뉴런 포함. 출력 뉴런은 제외)은 임시적으로 드롭아웃될 확률 p를 가짐
  - 이번 훈련 스텝에는 완전히 무시되지만 다음 스텝에서는 활성화될 수 있음
- 하이퍼파라미터 p를 드롭아웃 비율이라고 하고 보통 10%-50% 사이 지정
  - 순환 신경망에서는 20%-30%에 가까움
  - 합성곱 신경망에서는 40%-50%에 가까움
  - 훈련이 끝난 후에는 뉴런에 더는 드롭아웃을 적용하지 않음
- 드롭아웃으로 훈련된 뉴런은 이웃한 뉴런에 맞추어 적응될 수 없으므로 가능한 한 자기 자신이 유용해져야 함
  - 그래서 입력값의 작은 변화에 덜 민감해지고, 더 안정적인 네트워크가 되어 일반화 성능이 좋아짐
- TIP) 일반적으로 출력층을 제외한 맨 위의 층부터 세 번째 층까지 있는 뉴런에만 드롭아웃을 적용함
- p=50%로 하면 테스트하는 동안에는 하나의 뉴런이 훈련 때보다 평균적으로 두 배 많은 입력 뉴런과 연결됨
  - 훈련이 끝난 후 각 입력의 연결 가중치에 보존 확률(1-p)를 곱하거나
  - 훈련하는 동안 각 뉴런의 출력을 보존 확률로 나눌 수도 있음 (위의 방법과 동일하진 않지만 잘 동작함)
- 케라스에서는 `keras.layers.Dropout` 층을 사용하여 드롭아웃을 구현
- 주의) 드롭아웃은 훈련하는 동안에만 활성화되므로 훈련 손실과 검증 손실을 비교하면 안됨
비슷한 손실이더라도 훈련 세트에 과대적합될 수 있으므로 훈련이 끝난 후 드롭아웃을 빼고 훈련 손실을 평가해야함
- 모델이 과대적합되었다면 드롭아웃 비율을 늘리고, 반대로 과소적합되면 드롭아웃 비율을 낮춰야 함
- 층이 클 때는 드롭아웃 비율을 늘리고 작으 ㄴ층에는 드롭아웃 비율을 낮추는 것이 도움이 됨
  - 많은 최신 신경망 구조는 마지막 은닉층 뒤에만 드롭아웃을 사용함
- 드롭아웃은 수렴을 상당히 느리게 만드는 경향이 있지만 적절히 튜닝하면 훨씬 좋은 모델을 만들 수 있음
- TIP) 자기 정규화하는 네트워크를 규제하고 싶다면 알파 드롭아웃을 사용해야 함. 일반 드롭아웃은 자기 정규화하는 기능을 망가뜨릴 수 있음
